ml-agents/mlagents/trainers/policy/torch_policy.py

```diff
@@ -292,3 +261,6 @@ def init_load_weights(self) -> None:
 
     def get_weights(self) -> List[np.ndarray]:
         return []
+
+    def get_modules(self):
```

Comment:
I wonder if it could work with an implementation of `parameters()` to mimic the function of a `nn.Module`. Basically allowing non-nn.Modules to be saved and reloaded "like" a nn.Module

Reply:
Yes it's actually implemented that way with `state_dict()` in saver. So here I pass the modules and in saver I can simply call `state_dict()` and `load_state_dict()` on every module. This could be extended to non-nn.Modules by customizing their own `state_dict()` and `load_state_dict()` function. However I think a better practice would be making them all nn.Modules and putting everything in state_dict (will be easier once TF is deprecated).

Reply:
Not sure I would like to make all of them nn.Module. A lot of these are not modules and would not have a forward pass. 
What I am getting at is that instead of having the saver getting storing the nn.Modules, it could instead store a "savable" object (with save()/load() or even state_dict()/load_state_dict() to mimic nn.Modules). This way we could save things that are not necessarily modules (like GlobalSteps for instance).

Reply:
For now, as long as the object we passed in has state_dict()/load_state_dict(), they can be saved, not necessary being nn.Modules. I can change state_dict()/load_state_dict() to save()/load() and get_modules() to get_save_objects() to make it looks more general if that's better.
Yes the problem with GlobalSteps is that it does not live under nn.Module. Not sure in the future if we want to make policy or even trainer nn.Module so that all these things can go away with just one torch.save(state_dict()) / load_state_dict().

Reply:
Ok, I think I understand. 
Open question : Would it be super bad to replace `def get_modules(self)` with `def state_dict(self)`?
The `state_dict` method would get all the modules (including itself when needed) and call `state_dict` on all of them and combine them. Then it would return this "super" state_dict. We can also implement `load_state_dict` in a similar manner. This way, we can have the saver call `state_dict` and `load_dict` directly on the objects it want to save (even if they are not nn.Modules or multiple nn.Modules inside a single object) rather than collect all the nn.Modules.

Reply:
I get your idea of doing this recursively, but
1. Replacing the name to `state_dict(self)` is not a good idea since this function returns a different thing than `nn.Modules.state_dict()`. `get_modules()` returns the modules/savable objects themselves while `nn.Modules.state_dict()` returns the dictionary of layer names and weights. The saver needs the modules so that it can call `module.load_state_dict(state_dict)` on them. Naming these two with the same name would be very confusing.

2. My idea was leaving all the recursive work to nn.Modules. I can extend to non-nn.Modules if all objects has `get_modules()`, `.state_dict()` and `.load_state_dict()` functions. So it will be like this:
```
def get_modules(self):
        return {"Policy": self.actor_critic.get_modules(), "global_step": self.global_step.get_modules(), "some_other_object": self.some_object.get_modules()}
```
This function will return {"name": self} for nn.Modules and return custom dict for other objects. 
And then in Saver:
```
state_dict = {
            name: module.state_dict() for name, module in self.modules.items() # this is not the true torch state_dict() but some custom savable stuff for non-nn.Module objects.
}
torch.save(state_dict, f"{checkpoint_path}.pt")
```
, and similar for loading.

If it looks good I can do it. It's not hard to do but I'm a little bit concern how we make sure all the objects to be saved have this format (can't apply an abstract call for ALL objects apparently).

Reply:
Your comment convinced me, thanks for walking me through!

Reply:
With the current Saver framework, can it work with reward providers?

Reply:
I think so yes, we will add a get_module to the reward providers and have the optimizer update its module dictionary. I will take care of that in a later PR.